spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,541 @@
+    abortTransactionUnless(consensusStateAtProofHeight != nil)
+
+    // query the key/value proof on the indexed chain at the proof height
+    keyProof, _ := chain0.QueryProofAtHeight(key, proofHeight)
+
+    // assign the key/value proof
+    proof.KeyProof = &ProofData{
+        Key:   nil,  // key to prove constructed during verification
+        Value: nil,  // proven values are constructed during verification (except for frozen client proofs)
+        Proof: keyProof,
+    }
+
+    return &proof
+}
+
+// GenerateFrozenClientProof generate a multihop proof of a frozen channel given a set of chains starting from the misbehaving chain.
+// Chains are provided in order from the misbehaving chain to the verifying chain.
+func GenerateFrozenClientProof(chains []*Chain, proofHeight exported.Height) *MultihopProof {
+
+    abortTransactionUnless(len(chains) > 2)
+
+    // generate and assign consensus, clientState, and connection proofs
+    var proof MultihopProof
+    proof.ConsensusProofs = GenerateConsensusProofs(chains)
+    proof.ConnectionProofs = GenerateConnectionProofs(chains)
+
+    chain0 := chains[0] // misbehaving chain
+    chain1 := chains[1] // next hop chain
+
+    // check a consensus state at proof height is present on chain 1
+    consensusStateAtProofHeight := chain1.GetConsensusStateAtHeight(chain0, proofHeight)
+    abortTransactionUnless(consensusStateAtProofHeight != nil)
+
+    // connectionEnd on misbehaving chain representing the counterparty connection on the next chain
+    counterpartyConnectionEnd := abortTransactionUnless(Unmarshal(proof.ConnectionProofs[0].Value))
+
+    // key path to frozen client state
+    key := host.FullClientStatePath(counterpartyConnectionEnd.ClientId)
+    value := abortTransactionUnless(Marshal(chain1.GetClientState(counterpartyConnectionEnd.ClientId)))
+
+    // query the key/value proof on the next hop chain at the proof height
+    keyProof, _ := chain1.QueryProofAtHeight(key, proofHeight.GetRevisionHeight())
+
+    // assign the key/value proof
+    proof.KeyProof = &ProofData{
+        Key:   nil,    // key to prove constructed during verification using the provided connection state proofs
+        Value: value,  // provide the client state value
+        Proof: keyProof,
+    }
+
+    return &proof
+}
+
+
+// GenerateConsensusProofs generates consensus state proofs starting from the source chain to the N-1'th chain.
+// Compute proof for each chain 3-tuple starting from the source chain, C0 to the destination chain, CN
+// Step 1: |C0 ---> C1 ---> C2| ---> C3 ... ---> CN
+//         (i)    (i+1)    (i+2)
+// Step 2: C0 ---> |C1 ---> C2 ---> C3| ... ---> CN
+//                 (i)    (i+1)    (i+2)
+// Step N-3:  C0 ---> C1 ...  ---> |CN-2 --> CN-1 ---> CN|
+//                                   (i)    (i+1)    (i+2)
+// Note: It assumes that clients on  `C1,.., Cn` have been already updated in order.
```

Comment:
I think this algo will not work if some clients on path are updated in random order after this but before proof collection.